### PR TITLE
Support AWS ARM instances in Lokomotive

### PR DESCRIPTION
**Pull request by margamanterola**
# Support AWS ARM instances in Lokomotive

This change adds an os_arch parameter for both controller and workers on AWS.

### CODE REVIEW

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl [hunk 1]
```diff
@@ -141,15 +141,15 @@ storage:
       contents:
         inline: |
           KUBELET_IMAGE_URL=quay.io/kinvolk/kubelet
-          KUBELET_IMAGE_TAG=v1.19.4
+          KUBELET_IMAGE_TAG=v1.19.4-${os_arch}
```
Comment (invidian) [inline assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: This is not needed. We publish arm64 compatible manfiests:
```console
$ docker manifest inspect quay.io/kinvolk/kubelet:v1.19.4
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1994,
         "digest": "sha256:91f6f7fca8e5c3bc466d855496563e13ad3bc115106227bd46145e3d4910110d",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1994,
         "digest": "sha256:3d74bf03a3d088ec04374148b087d895f309f3ae31b89e758239f54a2c157996",
         "platform": {
            "architecture": "arm64",
            "os": "linux"
         }
      }
   ]
}
```

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl [hunk 2]
```diff
@@ -141,15 +141,15 @@ storage:
...
           NODE_LABELS="node.kubernetes.io/master,node.kubernetes.io/controller=true"
           NODE_TAINTS="node-role.kubernetes.io/master=:NoSchedule"
     - path: /etc/kubernetes/etcd.env
       filesystem: root
       mode: 0644
       contents:
         inline: |
-          IMAGE_TAG=v3.4.14
+          IMAGE_TAG=v3.4.14${etcd_arch_tag_suffix}
```
Comment (invidian) [inline assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:152]: etcd _should_ publish multiarch manifests from next release (https://github.com/etcd-io/etcd/pull/12630/commits/7f79ffb4092be1235d83a85657351145d3542405).

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl [hunk 3]
```diff
@@ -193,7 +194,7 @@ storage:
             -v /opt/bootkube/assets:/assets:ro \
             -v /etc/kubernetes:/etc/kubernetes:rw \
             --network=host \
-            quay.io/kinvolk/bootkube:v0.14.0-helm4 \
+            quay.io/kinvolk/bootkube:v0.14.0-helm4-${os_arch} \
```
Comment (invidian) [inline assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: our bootkube also offers multi-arch manifests.

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl [hunk 4]
```diff
@@ -170,6 +170,7 @@ storage:
           ETCD_PEER_CERT_FILE=/etc/ssl/etcd/etcd/peer.crt
           ETCD_PEER_KEY_FILE=/etc/ssl/etcd/etcd/peer.key
           ETCD_PEER_CLIENT_CERT_AUTH=true
+          ${etcd_arch_options}
```
Comment (invidian) [inline assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:173]: I think generic etcd_env_variables looping over string-string map would be better.

**Review by invidian — CHANGES_REQUESTED** [review]
We probably need a separate CI pipeline for AWS arm to be able to test components there.

File: pkg/platform/aws/aws.go [hunk 5]
```diff
@@ -35,6 +35,7 @@ type workerPool struct {
 	Count        int               `hcl:"count"`
 	SSHPubKeys   []string          `hcl:"ssh_pubkeys"`
 	InstanceType string            `hcl:"instance_type,optional"`
+	OSArch       string            `hcl:"os_arch,optional"`
```
Comment (invidian) [inline pkg/platform/aws/aws.go:38]: This must also be documented.

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl (hunk 1 above)
Comment (marga-kinvolk) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: I have no idea how this works :). I copied it from the packet module: https://github.com/kinvolk/lokomotive/blob/0b0c39c449f1b916182539f987de357e284b87a9/assets/terraform-modules/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl#L197

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl (hunk 3 above)
Comment (marga-kinvolk) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Ok, I've removed this and checked that indeed it downloads the right image. I guess it should be removed from the packet code?

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl (hunk 4 above)
Comment (marga-kinvolk) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: I've changed the name. I'm not sure what the string-string map would look like... Again, this is copied from the packet code and it probably shouldn't drift?

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl (hunk 1 above)
Comment (invidian) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Just remove the suffix and right image will be picked up automatically.

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl (hunk 3 above)
Comment (invidian) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Yes, but I guess it's out of scope, created an issue for it: #1379

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl (hunk 4 above)
Comment (invidian) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Something like:
```
%{ for key, val in etcd_env_variables ~}
${key}=${val}
%{ endfor ~}
```

>Again, this is copied from the packet code and it probably shouldn't drift?

Hmm... Either way is fine then I guess.

File: pkg/platform/aws/aws.go (hunk 5 above)
Comment (marga-kinvolk) [reply pkg/platform/aws/aws.go:38]: I've added the parameters to docs/configuration-reference/platforms/aws.md, is there another place I should document this?

Comment (invidian) [reply pkg/platform/aws/aws.go:38]: I think only `docs/configuration-reference/platforms/aws.md` is fine, thanks!

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl [hunk 6]
```diff
@@ -12,6 +12,20 @@ systemd:
...
+        Description=Set the host name from metadata service
+        Requires=coreos-metadata.service
+        After=coreos-metadata.service
+        [Service]
+        EnvironmentFile=/run/metadata/flatcar
+        ExecStart=/usr/bin/hostnamectl set-hostname $${COREOS_EC2_HOSTNAME}
+        RemainAfterExit=yes
+        Type=oneshot
+        [Install]
+        RequiredBy=wait-for-dns.service
```
Comment (invidian) [inline assets/terraform-modules/aws/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:28]: Why `wait-for-dns.service` requires it?

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl [hunk 7]
```diff
@@ -37,6 +37,20 @@ systemd:
       enable: true
     - name: locksmithd.service
       mask: true
+    - name: sethostname.service
```
Comment (invidian) [inline assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:40]: I think if we add this, we should have an e2e test for it to make sure hostname on matchine matches node name.

File: docs/configuration-reference/platforms/aws.md [hunk 8]
```diff
@@ -203,6 +203,7 @@ worker_pool "my-worker-pool" {
...
 | `cluster_name`                | Name of the cluster. **NOTE**: It must be unique per DNS Zone and region.                                                                                                                                                                                                                                    | -               | string       | true     |
 | `tags`                        | Optional details to tag on AWS resources.                                                                                                                                                                                                                                                                    | -               | map(string)  | false    |
+| `os_arch`                     | Flatcar Container Linux architecture to use (x86\_64, arm64).                                                                                                                                                                                                                                                | "x86\_64"        | string       | false    |
```
Comment (invidian) [inline docs/configuration-reference/platforms/aws.md:206]: Actually, maybe we can use https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/ec2_instance_type#supported_architectures to detect the architecture automagically using Terraform, based on instance type?

**Comment by surajssd** [comment]
@marga-kinvolk could you rebase and push again?

**Pull request closed without merging**